### PR TITLE
Add option to bundle app with pyinstaller

### LINKED ISSUE #486: Build portable single executable
Is it possible to build a single executable for simpler distribution of the tool?
For example with use of pyinstaller build a single `gcovr.exe` (for windows) file and use it in without need of 
installing python and pip?

Note: I think this is useful only for Windows platform.

### PR DESCRIPTION
**Pull request by Spacetown**
This PR adds an additional option to bundle the current version installed with pip into a standalone executable with pyinstaller:
```
Experimental Options:
  Experimental options are for testing only and can be changed/removed in
  future releases.

  --bundle-app          Build a standalone application from the current gcovr
                        installation. To add additional options to pyinstaller
                        use e.g.: --bundle-app -- --help
```

The main script used to generate the application is part of gcovr installation. In the bundled application the option and option group is removed.
The pipeline test run also a test for the bundled application which checks if the help is printed.

Closes #486

### CODE REVIEW

**Comment by latk** [comment]
I have no idea how PyInstaller works. Does this option have to be part of the program being packaged, or could it be a separate script, or could we just tell users interested in this to clone the repository and run some Nox command?

I think it's great to include (low-maintenance) entry-points that can be used for non-PyPI distribution, but the proposed design strikes me more like a hack, not like something I'd feel comfortable shipping.

**Comment by Spacetown** [comment]
Pyinstaller needs a script which is parsed to get the imports. For the modules the pyc files are created and packed together with the python interpreter. A bootstrap code unpacks all, setsup the environment and executes it.
Of course it can also be done in nox. Since this is something for more experienced users I thought a option would be better. But you're right, it's more like a little hack. I'll move it completely to the nox level and add it to the documentation. Should this go to cookbook?

**Comment by latk** [comment]
> I'll move it completely to the nox level and add it to the documentation. Should this go to cookbook?

Ok, from that description it really sounds like this should be a separate script, not part of the gcovr program. Doing this via Nox sounds fine.

Yes, adding it to the docs would be good for discoverability. Maybe a cookbook entry, and a link to that entry from the installation guide?

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -137,11 +137,6 @@ jobs:
         sudo apt-get clean
     - name: Install ninja
       uses: seanmiddleditch/gha-setup-ninja@master
-    - name: Install libxml2 and libxslt dev (pypy3)
-      if: ${{ startsWith(matrix.os,'ubuntu-') && ( matrix.python-version == 'pypy3' ) }}
-      run: |
-        sudo apt-get install -y libxml2-dev libxslt-dev
-        sudo apt-get clean
```
Comment (latk) [inline .github/workflows/test.yml:144]: Why was this deleted? If it isn't necessary, why was it added previously?

File: CHANGELOG.rst [hunk 2]
```diff
@@ -11,15 +11,20 @@ Known bugs:
 Breaking changes:
 
 New features and notable changes:
+
 - Accept `NAN %` which is used in GCOV 7.5.0 instead of an invalid value. (:issue:`651`)
 - New :option:`--json-base` to define a base bath used in JSON reports. (:issue:`656`)
+- New nox session `bundle_app` to generate a portable application with pyinstaller. (:issue:`661`)
```
Comment (latk) [inline CHANGELOG.rst:17]: Could link to the cookbook entry here.

Also, remember that RST needs double backticks for code markup :)

File: noxfile.py [hunk 3]
```diff
@@ -221,6 +222,63 @@ def upload_wheel(session: nox.Session) -> None:
...
+    """Bundle a standalone executable."""
+    session.install(
+        "jinja2",
+        "lxml",
+        "pygments",
+        "pyinstaller",
+    )
+    session.install("-e", ".")
+    if not os.path.exists("build"):
+        os.mkdir("build")
```
Comment (latk) [inline noxfile.py:236]: Can be simplified as:

```suggestion
    os.makedirs("build", exists_ok=True)
```

File: noxfile.py [hunk 4]
```diff
@@ -221,6 +222,63 @@ def upload_wheel(session: nox.Session) -> None:
...
+                """\
+                import sys
+                import runpy
+
+                if __name__ == "__main__":
+                    runpy.run_module("gcovr", run_name="__main__", alter_sys=True)
+                    sys.exit(0)
+                """
+            )
+        )
```
Comment (latk) [inline noxfile.py:251]: If we're creating a temporary file like this, we might as well add it directly as `scripts/pyinstaller_entrypoint.py` if that works.

File: noxfile.py (hunk 3 above)
Comment (latk) [inline noxfile.py:236]: Should add `/build/` to the .gitignore

File: noxfile.py [hunk 5]
```diff
@@ -221,6 +222,63 @@ def upload_wheel(session: nox.Session) -> None:
...
+@nox.session
+def bundle_app(session: nox.Session) -> None:
+    """Bundle a standalone executable."""
+    session.install(
+        "jinja2",
+        "lxml",
+        "pygments",
+        "pyinstaller",
+    )
+    session.install("-e", ".")
```
Comment (latk) [inline noxfile.py:234]: Is it necessary to install all these modules? I would guess that `install(".", "pyinstaller")` would be sufficient.

**Review by latk — COMMENTED** [review]
Ok, I like this. Lots of smaller questions though.

File: .github/workflows/test.yml (hunk 1 above)
Comment (Spacetown) [reply .github/workflows/test.yml:144]: While trying to get pyinstaller working on pypy I found out that this wasn't active since #550. With this PR pypy3 was replaced by pypy-3.7 which deactivates this step.

File: CHANGELOG.rst (hunk 2 above)
Comment (Spacetown) [reply CHANGELOG.rst:17]: 🙈

File: noxfile.py (hunk 3 above)
Comment (Spacetown) [reply noxfile.py]: Is already ignored. That's why I reused this folder.

File: noxfile.py (hunk 5 above)
Comment (Spacetown) [reply noxfile.py:234]: If I don't use the wheel I need to install all needed modules. On my computer, e.g. jinja isn't installed.

Comment (Spacetown) [reply noxfile.py:234]: The same as for the tests.

File: noxfile.py (hunk 4 above)
Comment (Spacetown) [reply noxfile.py:250]: This I need to do after my holiday. That's impossible on my mobile phone 😄

Comment (Spacetown) [reply noxfile.py:250]: If we use the wheel and there is an easy way to get the path to gcovr (independent of installation type global, local or venv) this will also work.

Comment (Spacetown) [reply noxfile.py]: @latk I've managed to use the script generated by setuptool. For Windows I need to find a way to extract the script from the executable.
